qiskit/circuit/quantumcircuit.py

```diff
@@ -2454,40 +2519,43 @@ def _append(self, instruction, qargs=(), cargs=()):
 
         :meta public:
         """
+        if _standard_gate:
+            new_param = self._data.append(instruction)
+            if new_param:
+                self._parameters = None
+            self.duration = None
+            self.unit = "dt"
+            return instruction
+
         old_style = not isinstance(instruction, CircuitInstruction)
         if old_style:
             instruction = CircuitInstruction(instruction, qargs, cargs)
-        self._data.append(instruction)
-        self._track_operation(instruction.operation)
-        return instruction.operation if old_style else instruction
+        params = None
+        # If there is a reference to the outer circuit in an
+        # instruction param we need to handle the params
+        # before calling the inner rust append method. This is to avoid trying
+        # to reference the circuit twice at the same time from rust. This shouldn't
+        # happen in practice but 2 tests were doing this and it's not explicitly
+        # prohibted by the API so this and the `params` optional argument path
+        # guard against it.
+        if hasattr(instruction.operation, "params") and any(
+            x is self for x in instruction.operation.params
+        ):
+            params = []
+            for idx, param in enumerate(instruction.operation.params):
+                if isinstance(param, (ParameterExpression, QuantumCircuit)):
+                    params.append((idx, list(set(param.parameters))))
+            new_param = self._data.append(instruction, params)
+        else:
+            new_param = self._data.append(instruction)
+        if new_param:
+            # clear cache if new parameter is added
+            self._parameters = None
```

Comment:
Can we sidestep this and do it with a `try_borrow` by accepting the right thing as a `PyCell` in Rust space?

Reply:
I can give this a try, it would be good to remove this awfulness from the append path. This ends up being one of the bigger impediments to performance in this PR in the circuit construction path

Reply:
I reworked this in: https://github.com/Qiskit/qiskit/pull/12459/commits/0281b6c9d7f4b1dd0e0ec87e3157ee3095e33d1c, I'm not sure it's exactly what you had in mind though.